flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcService.java

```diff
@@ -315,33 +315,48 @@ public void stopServer(RpcServer selfGateway) {
 			}
 
 			if (rpcEndpoint != null) {
-				akkaClient.getActorRef().tell(PoisonPill.getInstance(), ActorRef.noSender());
+				terminateAkkaRpcActor(akkaClient.getActorRef(), rpcEndpoint);
 			} else {
 				LOG.debug("RPC endpoint {} already stopped or from different RPC service", selfGateway.getAddress());
 			}
 		}
 	}
 
+	private CompletableFuture<Void> terminateAkkaRpcActor(ActorRef akkaRpcActorRef, RpcEndpoint rpcEndpoint) {
+		akkaRpcActorRef.tell(ControlMessages.TERMINATE, ActorRef.noSender());
+
+		return rpcEndpoint.getTerminationFuture();
+	}
+
 	@Override
 	public CompletableFuture<Void> stopService() {
+		LOG.info("Stopping Akka RPC service.");
+
+		final CompletableFuture<Void> akkaRpcActorsTerminationFuture;
+
 		synchronized (lock) {
 			if (stopped) {
 				return terminationFuture;
 			}
 
 			stopped = true;
-		}
 
-		LOG.info("Stopping Akka RPC service.");
+			final Collection<CompletableFuture<Void>> akkaRpcActorTerminationFutures = new ArrayList<>(actors.size());
```

Comment:
The section that starts here could go into a private method like `CompletableFuture<Void> terminateAllActors()`

Reply:
Good point, will extract a method here.